File: src/main/java/org/opendsb/messaging/DataMessage.java
package org.opendsb.messaging;

import org.opendsb.json.AnnotationExclusion;
import org.opendsb.json.TypedData;
import org.opendsb.json.info.DefaultData;

import com.google.gson.Gson;
import com.google.gson.GsonBuilder;

public class DataMessage extends BaseMessage {
	
	private static Gson coder = new GsonBuilder().setExclusionStrategies(new AnnotationExclusion())
			.registerTypeAdapter(TypedData.class, new TypedData.TypedDataAdapter()).create();

	private TypedData data;

	public DataMessage(String destination, String origin, Object data) {
		super(destination, origin);
		this.type = MessageType.PUBLISH;
		
		if (data instanceof TypedData) {
			this.data = (TypedData)data;
		} else {
			this.data = new DefaultData(data);
		}
	}

	public Object getData() {
		return data.getData();
	}

	public static <T extends DataMessage> T fromJSON(String json, Class<T> clazz) {
		return coder.fromJson(json, clazz);
	}
	
	@Override
	public String toJSON() {
		return coder.toJson(this);
	}
	
}

File: python/src/opendsb/messaging/datamessage.py

from abc import ABC
import json
from .basemessage import BaseMessage
from .message import MessageType


class TypedData(ABC):
    def __init__(self, data, concreteType: str) -> None:
        self.data = data
        self.concreteType = concreteType

class DefaultData(TypedData):
    def __init__(self, data: dict) -> None:
        super().__init__(data['payLoad'], 'org.opendsb.json.info.DefaultData')
        self.dataType = data['dataType']

    def to_dict(self):
        return {
            'data': self.data,
            'dataType': self.dataType,
            'concreteType': self.concreteType
        }

class TypedCollection(TypedData):
    def __init__(self, data: dict) -> None:
        super().__init__(data['payLoad'], 'org.opendsb.json.info.TypedCollection')
        self.collectionRawType = data['collectionRawType'] # tipo da colecao em si
        self.collectionGenericType = data['dataType'] # tipo do conteudo da colecao


class DataMessage(BaseMessage):
    def __init__(self, 
                 destination: str, 
                 origin: str, 
                 data: str | dict | list | int | float | bool | None,
                 id: str | None = None,
                 latest_hop: str | None = None,
                 type: MessageType = MessageType.PUBLISH
        ):
        super().__init__(origin=origin, destination=destination, type=type, id=id, latest_hop=latest_hop)
        self.data = data

    def __repr__(self) -> str:
        attributes = vars(self)
        formatted_attributes = [f'{var}={value}' for var, value in attributes.items()]
        return f"{type(self).__name__}({', '.join(formatted_attributes)})"
   
    @classmethod    
    def from_json(cls, json_message: str):
        #message_dict = json.loads(json_message, object_hook=cls.flatten_hook)
        message_dict = json.loads(json_message)
        ATTR_MAP_INV = {v: k for k, v in cls.ATTR_MAP.items()}
        message_dict = {ATTR_MAP_INV[attr]: value for attr, value in message_dict.items()}
        return cls(message_dict['destination'], message_dict['origin'], message_dict['data'], message_dict['id'], message_dict['latest_hop'])

    